conduit/pipeline/pipeline.go

```diff
@@ -431,8 +467,250 @@ func addMetrics(block data.BlockData, importTime time.Duration) {
+			}
+		}()
+
+		for {
+			p.logger.Tracef("exporter handler waiting for block data. round=%d", lastRound)
+			selectStart := time.Now()
+			select {
+			case <-p.ctx.Done():
+				p.logger.Infof("exporter handler exiting lastRnd=%d", lastRound)
+				return
+			case pluginBlk := <-blkChan:
+				waitTime := time.Since(selectStart)
+				totalSelectWait += waitTime
+				p.logger.Tracef("exporter handler received block data for round %d after wait of %s", lastRound, waitTime)
+				lastRound = pluginBlk.Round()
+
+				if p.pipelineMetadata.NextRound != lastRound {
+					lastError = fmt.Errorf("aborting after out of order block data. %d != %d", p.pipelineMetadata.NextRound, lastRound)
+					return
+				}
+
+				var exportTime time.Duration
+				exportTime, lastError = RetriesNoOutput(exporter.Receive, pluginBlk.BlockData, p, exporter.Metadata().Name)
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after failing to export round %d: %w", lastRound, lastError)
+					return
+				}
+				metrics.ExporterTimeSeconds.Observe(exportTime.Seconds())
+				// Ignore round 0 (which is empty).
+				if lastRound > 0 {
+					// Previously we reported time starting after block fetching is complete
+					// through the end of the export operation. Now that each plugin is running
+					// in its own goroutine, report only the time of the export.
+					addMetrics(pluginBlk.BlockData, exportTime)
+				}
+
+				// Increment Round, update metadata
+				nextRound := lastRound + 1
+				p.pipelineMetadata.NextRound = nextRound
+				_, lastError = RetriesNoOutput(p.pipelineMetadata.encodeToFile, p.cfg.ConduitArgs.ConduitDataDir, p, "pipelineMetadata save")
+				if lastError != nil {
+					lastError = fmt.Errorf("aborting after updating NextRound=%d and failing to save metadata: %w", nextRound, lastError)
+					return
+				}
+				p.logger.Tracef("exporter %s @ round %d saved pipeline metadata", exporter.Metadata().Name, p.pipelineMetadata.NextRound)
+
+				for i, cb := range p.completeCallback {
+					p.logger.Tracef("exporter %s @ round=%d NextRound=%d executing callback %d", exporter.Metadata().Name, lastRound, p.pipelineMetadata.NextRound, i)
+					_, lastError = RetriesNoOutput(cb, pluginBlk.BlockData, p, fmt.Sprintf("callback %d", i))
+					if lastError != nil {
+						lastError = fmt.Errorf("aborting due to failed callback %d: %w", i, lastError)
+						return
+					}
+				}
+				lastError = nil
+				// WARNING: removing the following will BREAK:
+				// - the E2E test (Search for "Pipeline round:" in subslurp.py)
+				// - the internal tools logstats collector (See func ConduitCollector in logstats.go of internal-tools repo)
+				// Modify with CAUTION!!!!
+				p.logger.Infof(
+					"UPDATED Pipeline NextRound=%d after [%s] from round kickoff. FINISHED Pipeline round r=%d (%d txn) exported in %s",
+					nextRound,
+					time.Since(pluginBlk.startRoundTime),
+					lastRound,
+					len(pluginBlk.Payset),
+					time.Since(pluginBlk.finishImportTime),
+				)
```

Comment:
If we take the end of round work out of the exporter, then we'll need a return channel -so that dovetails with your comment. But if we want to keep a semantically equivalent log message (for use by internal tools `logstats.go`) the timer needs to start at the end of the import, and needs to be sent down through the channels so I don't see how to avoid something like this.

In my current EC2 testing, I haven't found the extra log information very useful. In fact, it can be misleading because there is work happening concurrently so you can't just add up the processing times and figure out the contribution to the overall time.

My proposal is:
1. revert to channels of type `chan data.BlockData`
2. exporter provides a return channel of this type
3. roundFinishHandler reads the channel. In particular it:
   * does each portion of the work _failing fast without any retries_
   * logs the "...FINISHED Pipeline round r=..." message
4. modify `logstats.go` so that it _doesn't care_ about the time to export
5. we can add more TRACE logs to get precise metrics if we need in the future

Reply:
There is a more ambitious approach which is to have a metrics collector keep track of all the times each plugin starts/stops and then do its own calculations accordingly. But that seems like too much of a rabbit hole to go into at this juncture.

Reply:
I'm proposing only recording the "exporter". Unfortunately, when the importer < exporter, under pipelining we can no longer deduce the "importer" time. However, a future possibility is to for the importer to add a separate log.

Reply:
867973f82e586f7d736e53e44eaf6d1f08871275